src/lpython/parser/tokenizer.re

```diff
@@ -426,6 +426,7 @@ int Tokenizer::lex(Allocator &al, YYSTYPE &yylval, Location &loc, diag::Diagnost
             'not'  { RET(TK_NOT) }
             'and'  { RET(TK_AND) }
             'or'   { RET(TK_OR) }
+            'is' whitespace 'not'  { RET(TK_IS_NOT) }
```

Comment:
I think instead of using whitespace; we should use something generic.

Here is the issue with using white space -

```
❯ cat examples/test1.py
x = 10
if x is \
    not type(int):
    pass
```
Old Parser
```
❯ lpyast examples/test1.py --indent
(Module [
   (Assign [
      (Name
         x
         Store)]
      (ConstantInt 10 ()) ())
   (If
      (Compare
         (Name
            x
            Load)
         IsNot [
         (Call
            (Name
               type
               Load) [
            (Name
               int
               Load)] [])]) [
      (Pass)] [])] [])
```
New Parser
```
❯ lpynewpar examples/test1.py --indent
(Module [
   (Assign [
      (Name
         x
         Store)]
      (ConstantInt 10 ()) ())
   (If
      (Compare
         (Name
            x
            Load)
         Is [
         (UnaryOp
            Not
            (Call
               (Name
                  type
                  Load) [
               (Name
                  int
                  Load)] []))]) [
      (Pass)] [])] [])
```
Diff
```diff
❯ diff old.txt new.txt
12,19c12,21
<          IsNot [
<          (Call
<             (Name
<                type
<                Load) [
<             (Name
<                int
<                Load)] [])]) [
---
>          Is [
>          (UnaryOp
>             Not
>             (Call
>                (Name
>                   type
>                   Load) [
>                (Name
>                   int
>                   Load)] []))]) [
```

Reply:
Yup, We have to make it. But, for now we can do something like this:
```diff
diff --git a/src/lpython/parser/tokenizer.re b/src/lpython/parser/tokenizer.
re
index 54449d338..bda807106 100644
--- a/src/lpython/parser/tokenizer.re
+++ b/src/lpython/parser/tokenizer.re
@@ -367,7 +367,6 @@ int Tokenizer::lex(Allocator &al, YYSTYPE &yylval, Locat
ion &loc, diag::Diagnost
                 RET(TK_NEWLINE);
             }
 
-            "\\" newline { continue; }
 
             // Single character symbols
             "(" { token_loc(loc); record_paren(loc, '('); RET(TK_LPAREN) }
@@ -427,6 +426,9 @@ int Tokenizer::lex(Allocator &al, YYSTYPE &yylval, Locat
ion &loc, diag::Diagnost
             'and'  { RET(TK_AND) }
             'or'   { RET(TK_OR) }
             'is' whitespace 'not'  { RET(TK_IS_NOT) }
+            'is' whitespace? "\\" newline whitespace? 'not'  { RET(TK_IS_NO
T) }
+
+            "\\" newline { continue; }
 
             // True/False
```
@certik, what are your thoughts?

Reply:
I would start with your latest diff, add tests. That should cover 99% of cases. Then later we can revisit this. For example, once we can parse all of Python, we can see if there is a way to parse this in Bison. It's too early now, as we will be adding more rules, and so it's safer to do it from the tokenizer. We can also later investigate if there is a way to make this 100% robust in the tokenizer.